RecoLocalTracker/SiPixelRecHits/plugins/alpaka/SiPixelRecHitExtendedAlpaka.cc

```diff
@@ -0,0 +1,198 @@
+    // this code merges and copy both of them into a new SoA collection
+    // taking into account that for the hits the copy is straightforward,
+    // while for the hitModuleSoA we need to copy nPixelModules + nTrackerModules + 1 elements
+    // to account for the last "hidden" element in the SoA which is used to store
+    // the cumulative sum of hits in all the previous modules (thus this SoA has 1 more
+    // element than the actual number of modules to track the hits in the last module
+    // and sum them to the others).
+    // See also DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsDevice.h
+
+    const int nPixHits = pixColl.nHits();
+    const int nTrkHits = trkColl.nHits();
+
+    const int nPixMod = pixColl.nModules();
+    const int nTrkMod = trkColl.nModules() + 1;  // +1 for the "hidden" last element in the SoA
+
+    // the output is also a SoA collection with the same layout as the input ones
+    auto output = reco::TrackingRecHitsSoACollection(queue, nPixHits + nTrkHits, nPixMod + nTrkMod);
+
+#ifdef GPU_DEBUG
+    std::cout << "----------------- Merging Pixel and Tracker RecHits -----------------\n"
+              << "Number of Pixel recHits: " << nPixHits << '\n'
+              << "Number of Tracker recHits: " << nTrkHits << '\n'
+              << "Total number of recHits: " << output.nHits() << '\n'
+              << "Number of Pixel modules: " << nPixMod << '\n'
+              << "Number of Tracker modules: " << nTrkMod - 1 << '\n'
+              << "Total number of modules: " << output.nModules() << '\n'
+              << "---------------------------------------------------------------------\n";
+#endif
+
+    // start from the hits SoA, use metarecords to loop over all the columns
+    auto outView = output.view();
+    auto pixView = pixColl.view();
+    auto trkView = trkColl.view();
+
+    // layout type (same for all views)
+    using ViewType = decltype(outView);
+    using LayoutType = typename ViewType::Metadata::TypeOf_Layout;
+
+    // build descriptors (tuple of spans: one span for each column)
+    auto outDesc = LayoutType::Descriptor(outView);
+    auto pixDesc = LayoutType::ConstDescriptor(pixView);
+    auto trkDesc = LayoutType::ConstDescriptor(trkView);
+
+    // merge all columns using a compile-time loop
+
+    // number of columns (same for all hits SoAs)
+    constexpr std::size_t N = std::tuple_size_v<decltype(outDesc.buff)>;
+
+    mergeSoAColumns<N>([&](auto columnIndex) {
+      auto& outCol = std::get<columnIndex>(outDesc.buff);
+      const auto& pixCol = std::get<columnIndex>(pixDesc.buff);
+      const auto& trkCol = std::get<columnIndex>(trkDesc.buff);
+      // FIXME offsetBPIX2 is the only scalar column, all others are arrays
+      // and should be copied as such, I have not found a better way to distinguish it from the others,
+      // the code below is a workaround to copy the scalar column (as long as the layout does not change)
+      if constexpr (columnIndex == 13) {
```

Comment:
This approach is not maintainable.
@leobeltra is there a way to check the properties of a SoA entry, like if it's a column or scalar ?

Reply:
Thanks for the insight. Is there a way to do the same thing, but accessing the column by index instead of name? (the check is done in a compile-time loop over the columns where I don't necessarily have access to their name, or view)

Reply:
(possibly unrelated, but I think you should use `std::string_view` instead of `const char*` to return the SoA entry name)

Reply:
As a side note, this check could also be size-based since I assume `column.size()` would return 1 for a scalar. I'm open to suggestions on what the best implementation would be